presto-main/src/main/java/com/facebook/presto/operator/aggregation/TDigestAggregationFunction.java

```diff
@@ -0,0 +1,115 @@
+
+public class TDigestAggregationFunction
+        extends StatisticalDigestAggregationFunction
+{
+    public static final TDigestAggregationFunction TDIGEST_AGG = new TDigestAggregationFunction(parseTypeSignature("V"));
+    public static final TDigestAggregationFunction TDIGEST_AGG_WITH_WEIGHT = new TDigestAggregationFunction(parseTypeSignature("V"), parseTypeSignature(StandardTypes.BIGINT));
+    public static final TDigestAggregationFunction TDIGEST_AGG_WITH_WEIGHT_AND_COMPRESSION = new TDigestAggregationFunction(
+            parseTypeSignature("V"),
+            parseTypeSignature(StandardTypes.BIGINT),
+            parseTypeSignature(StandardTypes.DOUBLE));
+
+    private static final MethodHandle INPUT_DOUBLE = methodHandle(TDigestAggregationFunction.class, "inputDouble", StatisticalDigestState.class, double.class, long.class, double.class);
+
+    public static final String NAME = "tdigest_agg";
+
+    private TDigestAggregationFunction(TypeSignature... typeSignatures)
+    {
+        super(NAME, TDIGEST, createTDigestFactory(), typeSignatures);
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Returns a tdigest from the set of doubles";
+    }
+
+    public static void inputDouble(StatisticalDigestState state, double value, long weight, double compression)
+    {
+        checkArgument(weight > 0, "Weight must be > 0, was %s", weight);
+        checkArgument(compression > 0, "Compression factor must be positive, was %s", compression);
+
+        StatisticalDigest digest = getOrCreateTDigest(state, compression);
+        state.addMemoryUsage(-digest.estimatedInMemorySizeInBytes());
+        digest.add(value, weight);
+        state.addMemoryUsage(digest.estimatedInMemorySizeInBytes());
+    }
+
+    private static StatisticalDigest getOrCreateTDigest(StatisticalDigestState state, double compression)
+    {
+        StatisticalDigest digest = state.getStatisticalDigest();
+        if (digest == null) {
+            digest = new StatisticalTDigest(createTDigest(compression));
+            state.setStatisticalDigest(digest);
+            state.addMemoryUsage(digest.estimatedInMemorySizeInBytes());
+        }
+        return digest;
+    }
+
+    protected MethodHandle getInputMethodHandle(Type valueType, int arity)
+    {
+        final MethodHandle inputFunction;
+        switch (valueType.getDisplayName()) {
+            case StandardTypes.DOUBLE:
+                inputFunction = INPUT_DOUBLE;
+                break;
+            case StandardTypes.REAL:
+                throw new PrestoException(NOT_SUPPORTED, "Cannot operate on a t-digest with real numbers");
+            case StandardTypes.BIGINT:
+                throw new PrestoException(NOT_SUPPORTED, "Cannot operate on a t-digest with longs");
+            default:
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Unsupported type %s supplied", valueType.getDisplayName()));
+        }
+
+        switch (arity) {
```

Comment:
Is this part of logic also the same in both implementations? If `specialize` is in the abstract class, `arity` should also be handled there.

Reply:
Just moved the `getInputMethodHandle()` functions into the abstract class. It does make `StatisticalDigestAggregationFunction` look messier though. You can see the changes in this gist:
<script src="https://gist.github.com/gastonar/2f6f54c394750802000b33ba899768b4.js"></script>